internal/controller/bucket_controller.go

```diff
@@ -943,3 +749,162 @@ func fetchIndexFiles(ctx context.Context, provider BucketProvider, obj *sourcev1
 
 	return nil
 }
+
+// setupCredentials retrieves and validates secrets for authentication, TLS configuration, and proxy settings.
+// It returns the primary secret, proxy URL, TLS config, STS secret, and STS TLS config.
+func (r *BucketReconciler) setupCredentials(ctx context.Context, obj *sourcev1.Bucket) (*corev1.Secret, *url.URL, *tls.Config, *corev1.Secret, *tls.Config, error) {
```

Comment:
I would create a struct `BucketCredentials` and use it here like so
```suggestion
func (r *BucketReconciler) setupCredentials(ctx context.Context, obj *sourcev1.Bucket) (*BucketCredentials, error) {
```

Reply:
Thanks for the suggestion. 
I didn't export the struct because credentials for buckets aren't meant to be used outside the package. 
They're only used as args in unexported functions, so I named it `bucketCredentials`.

Reply:
I’ve also applied the same approach to the HelmRepository.

[51cb0e0](https://github.com/fluxcd/source-controller/pull/1852/commits/51cb0e03b40712edef5957636a60669ac1fa2c0e)